### ISSUE TITLE
tensor `.data` assignment fails in PyTorch XLA when shapes are different (unlike on CPUs)

### ISSUE THREAD

**Issue by ronghanghu** [pytorch/xla#3330]
## 🐛 Bug

One cannot assign a different tensor to a tensor's `.data` attribute in PyTorch XLA due to its IR tracing. Instead of pointing a tensor's `.data` handle to a new memory blob, this assignment would cause the code to crash. (This is unlike in the vanilla PyTorch on CPU or GPU, where one can do it without errors.)

(Is this an expected behavior in PyTorch XLA or a bug?)

## To Reproduce

The following code that tries to assign a tensor's `.data` to another tensor of a different shape would crash when running on TPUs
```
import torch
import torch_xla.core.xla_model as xm

device = xm.xla_device()

p = torch.zeros(21904, device=device)
tmp1 = p.data[:21900]  # this is OK
tmp2 = tmp1.view(21900, 1)  # this is OK
p.data = tmp2  # this crashes in XLA
```

giving the following error
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: torch_xla/csrc/tensor.cpp:781 : Check failed: xla::util::Multiply<xla::int64_t>(ir_value.shape().dimensions()) == xla::util::Multiply<xla::int64_t>(view->shape().dimensions()) (21900 vs. 21904)
*** Begin stack trace ***
        tensorflow::CurrentStackTrace()
        torch_xla::XLATensor::UpdateView(std::shared_ptr<torch_xla::View>, torch_xla::ir::Value) const
        torch_xla::XLATensor::SetIrValue(torch_xla::ir::Value)
        torch_xla::XLATensor::ShallowCopyTo(torch_xla::XLATensor*) const
        torch_xla::XLATensorImpl::shallow_copy_from(c10::intrusive_ptr<c10::TensorImpl, c10::detail::intrusive_target_default_null_type<c10::TensorImpl> > const&)

        THPVariable_set_data(THPVariable*, _object*, void*)

        PyObject_SetAttr
        _PyEval_EvalFrameDefault
        _PyEval_EvalCodeWithName
        PyEval_EvalCode



        PyRun_InteractiveLoopFlags
        PyRun_AnyFileExFlags

        Py_BytesMain
        __libc_start_main
        _start
*** End stack trace ***
```

## Expected behavior

On CPU (or GPU), the same code would work and one can assign a tensor of a different shape to another tensor's `.data`.
```
import torch

device = torch.device("cpu")

p = torch.zeros(21904, device=device)
tmp1 = p.data[:21900]  # this is OK
tmp2 = tmp1.view(21900, 1)  # this is OK
p.data = tmp2  # this is OK on CPU
```

## Environment

 - Reproducible on XLA backend [CPU/TPU]: v3-8 TPU VM (tpu-vm-pt-1.10)
 - torch_xla version: 1.10

## Additional context

This error emerges when I was trying to implement the ZeRO optimizer in PyTorch XLA and needed to change a parameter tensor's `.data` to point to a sharded and sliced tensor of a smaller size. If this is the expected behavior in PyTorch XLA, I can also try to find a workaround to it.

**Comment by JackCaoG** [comment]
Thanks @ronghanghu for reporting, I will take a look

**Comment by ronghanghu** [comment]
@JackCaoG Upon further look, it seems that the error is because PyTorch XLA's IR tracing does not allow assigning a tensor's `.data` to a slice of itself (although it still allows assigning its `.data` to another tensor of a different shape).

A simpler snippet to reproduce the error is as follows
```
import torch
import torch_xla.core.xla_model as xm

device = xm.xla_device()

p = torch.zeros(21904, device=device)
p.data = torch.zeros(21900, device=device)  # this is OK on XLA
p.data = p.data[:20000]  # crashes on XLA (but works on CPU)
```

The last slicing step comes from my ZeRO implementation where I try to trim the padding dimension from a parameter tensor (generated via all-gathering of its shards). I can also find another workaround to implement it.

**Comment by ronghanghu** [comment]
Note that on the other hand, in-place slicing assignment could work for a tensor's `.grad` as below, which makes it further weird that it won't work for a tensor's `.data`

```
import torch
import torch_xla.core.xla_model as xm

device = xm.xla_device()

p = torch.zeros(8000, device=device, requires_grad=True)
p.sum().backward()

p.data = torch.zeros(1000, device=device)
assert p.grad.numel() == 8000
p.grad = p.grad[:1000]  # this is OK on XLA without crashing
```

**Comment by JackCaoG** [comment]
Ah, OK, this is an interesting problem. @bdhirsh since he implemented the `UpdateView`

I used
```
# this is OK on XLA
t1 = torch.zeros(80, device=device)
t1.data = torch.zeros(10, device=device)

# all of the following crash on XLA (but work on CPU)
t1 = torch.zeros(80, device=device)
t1_slice = t1.data[:10]
t1.data = t1_slice
```
to repo and it crashed on `t1.data = t1_slice`. This only happens when you are trying to assign the the value to a tensor's data when it has a view tensor associated with it.

What happened is that `t1_slice` is a view of `t1`. When we try to assign something to `t1.data`, pt'/xla knows there is a view associated with `t1`. For the correctness when we update `t1`'s data, we need to update all of the view that is associated with it. It is a long story getting into how pytorch/xla implement view (since we don't really have storage), but we pretty much remembers what kind of `reshape` operation we performed on each view and replay all of the update after the reshape. As you can image, that would require the original base view base tensor to have the same shape or at least have the same number of the elements. Otherwise reshape operation will not make sense.

Even if you do `.clone().detach`, resulting tensor is still a view of the base tensor
```
(Pdb) t1_slice = t1.data[:10].clone().detach()
(Pdb) print(torch_xla._XLAC._get_xla_tensors_text([t1_slice]))
IR {
  %0 = f32[] prim::Constant(), location=<module>@test.py:56, value=0
  %1 = f32[80]{0} aten::expand(%0), location=<module>@test.py:56, size=(80)
  %2 = f32[10]{0} xla::select(%1), location=<module>@<stdin>:1, dim=0, start=0, end=10, stride=1, ROOT=0
}
```

I guess the workaround is to extract part of the tensor without using a view operation. @bdhirsh does such op exists?

**Comment by JackCaoG** [comment]
> I'm surprised that in your example though with `t1_slice = t1.data[:10].clone().detach()`, that the output is still a view. Since I'd expect `clone()` to make a copy of the current tensor, and the resulting tensor that it creates to not have any view information.

When doing a copy we certainly don't want to materialize the ir of the base tensor so we pretty much just do a copy of the IR. This is usually fine but when the tensor we want to clone is a view tensor then pt/xla behavior diverge from the pytorch. One possibility is when a `clone` is issued, we make a copy of the base tensor and rebuild the IR based on the new copied base tensor. It sounds complicated and I am not sure if this issue will be fixed by your upcoming functionialzation pass.

**Comment by ronghanghu** [comment]
@JackCaoG @bdhirsh Upon further inspections, I found that the following code snippet using `torch.split` would allow me to get around the issue above and assign a tensor's `.data` to its slice.
```
import torch
import torch_xla.core.xla_model as xm

device = xm.xla_device()

# this is OK on XLA using `torch.split`
p = torch.zeros(80, device=device)
p.data, _ = torch.split(p.data, [10, 70])

# # this crashes on XLA (but works on CPU)
# p = torch.zeros(80, device="device")
# p.data = p.data[:10]
```

The trick of ``torch.split` also works for me in practical use cases beyond the toy example above. However, it is a little confusing as according to the [documentation](https://pytorch.org/docs/stable/generated/torch.split.html) of `torch.split` says "Each chunk is a view of the original tensor". @JackCaoG Is this the intended behavior in PyTorch XLA? And why would `torch.split` be able to get around the slice issue despite that it should also generate views?

**Comment by JackCaoG** [comment]
@ronghanghu I checked our code and we did not lower split as a view op. This is where we are misaligned with the pytorch behavior. I tried to modify the result of `split` and it didn't update the base tensor.

I checked the pr https://github.com/pytorch/xla/commit/3d1bc39624d7e493c49ec403e99c97267392f94f and it is one of those earlier ops we lowered so my guess is either we forgot to make it a view op, we did not have a mechanism to support the view op in the earliest day or `split` was initially not a view op but behavior change between it was lowered and now. After @bdhirsh's work on funcitonization pass is upstream and we adapt to that, it should fairly easy for us to support all view ops. @bdhirsh Is there a quick way for us to check all of the view ops? I suspect `split` is not the only one.

@ronghanghu Do you felt like making `split` a view op is an important thing you need us to fix ASAP? IMO making a op to be view op can usually save times on CPU and CUDA backend because it avoid additional data allocation. However it is more expensive for the PyTorch/XLA and other backends doesn't expose storage to the front end.

**Comment by bdhirsh** [comment]
@JackCaoG yep, the pass today generates a list of "non-composite aliasing ops" that a backend would need to write _copy() variants of. This list is pretty long because it includes *every* (non-composite) aliasing op, including some that aren't very common and/or are private ops, but for completeness:
```
_fw_primal
_make_dual
view_as_real
view_as_complex
_conj
_neg_view
as_strided
_sparse_broadcast_to
diagonal
expand
permute
_reshape_alias
select_int
detach
slice_Tensor
split_Tensor
split_with_sizes
squeeze
squeeze_dim
t
transpose_int
unsqueeze
_indices
_values
indices
values
crow_indices
col_indices
unbind_int
view
view_dtype
unfold
alias
```

**Comment by JackCaoG** [comment]
@bdhirsh I felt like it would be much easier to fix after funcitonization pass, wdyt?

**Comment by ronghanghu** [comment]
@JackCaoG Thanks for the explanation!

> @ronghanghu Do you felt like making split a view op is an important thing you need us to fix ASAP? IMO making a op to be view op can usually save times on CPU and CUDA backend because it avoid additional data allocation. However it is more expensive for the PyTorch/XLA and other backends doesn't expose storage to the front end.

I think the current XLA implementation of `torch.split` works for me in training, but I'm still trying to understand its memory implication. I would like to explain my usage as follows and hope to get some insights from you.

---

My actual use case is the ZeRO optimizer, where I'm trying to concatenate all the parameters in a `nn.Module` layer into a single vector for sharding, and then splitting it to get individual parameters as follows.

```
# implementation a layer for ZeRO-3
def forward(self, inputs):
    ...
    # Step 1. gather a concatenated version of layer params from per-TPU shards
    full_concat_param = all_gather(self.sharded_concat_param)

    # Step 2. split the concatenated version into actual parameters
    # - question: do p1, p2, p3 occupy the same storage as full_concat_param?
    #   e.g. if `full_concat_param` is 2 GB size, would this op take 2 GB or 4 GB storage on TPU memory?
    self.p1.data, self.p2.data, self.p3.data = torch.split(full_concat_param, [p1_dim, p2_dim, p3_dim])

    # Step 3. use the actual param to perform forward computation
    outputs = self.inner_module(inputs, self.p1, self.p2, self.p3)

    # Step 4. free the storage of params after forward pass is done for this layer (we'll gather again before backward pass)
    # - question: does it actually free storage of full_concat_param or p1, p2, p3?
    #    e.g. can the next layer's forward pass reuse the TPU memory occupied by `full_concat_param`?
    #    if not, is there a way to explicitly destroy the buffer `full_concat_param` to free the storage?
    #    like modifying a tensor's `.storage` in CUDA?
    full_concat_param.data = full_concat_param.data.new_zeros(1)

...
def pre_backward_hook(self, ...):
    ...
    # Step 5. reconstruct the parameters again for backward pass
    full_concat_param = all_gather(self.sharded_concat_param)
    self.p1.data, self.p2.data, self.p3.data = torch.split(full_concat_param, [p1_dim, p2_dim, p3_dim])
```

Question 1: In my ZeRO-3 use case above, if I generate split tensors (my actual parameters) from a tensor using `p1, p2, p3 = torch.split(full_concat_param, ...)`, *would it reuse the same TPU memory (which would be better and more memory efficient) or would it need to actually allocate new memory?* (e.g. if `full_concat_param` is 2 GB, would this layer's forward pass take 2 GB or 4 GB storage on TPU memory via `torch.split`?)

Question 2: Another memory-related question is that in ZeRO-3 implementation, I hope to free the memory storage of the gathered parameter tensors in one layer after its forward pass is done (so that the next layer's forward can reuse the same TPU memory) while I need to keep the `.data` handles of parameter tensors so that PyTorch autograd can work on them (their parameters will be gathered again into the same tensor `.data` during the layer's backward pass). 

My current implementation is to call something like `t.data = t.data.new_zeros(1)` in an attempt to free storage after the forward pass and reconstruct them later in a hook before the backward pass. *Does this allow me to actually free a parameter tensor's memory on TPU? If not, is there a way to explicitly destroy the buffer `full_concat_param` to free the storage?*

**Comment by JackCaoG** [comment]
Let me use this example to answer your question
```
(pytorch2) jackcao@jackcao64:~/Desktop/gitProj2/pytorch/xla$ cat ~/Desktop/test/test_view.py
import torch
import torch_xla
import torch.nn as nn
import torch_xla.core.xla_model as xm
import torch_xla.debug.metrics as met


device = xm.xla_device()

#  each int takes 4 byte
t1 = torch.zeros(40, device=device)
t2 = torch.zeros(20, device=device)
t3 = torch.zeros(20, device=device)
import pdb; pdb.set_trace()
t2.data, t3.data = torch.split(t1.data, [20, 20])

# allocate 320 b memroy
xm.mark_step()

# free t1's allocation, 160 b
t1 = None
# free t2's allocation, 80 b
t2 = None
# free t3's allcation, 80 b
t3 = None

# print 3, since we destory 3 data handle
print(met.metric_data('ReleaseDataHandlesTime'))
```

In short, split will allocate new memory. Setting the tensor value to None(and make sure this is the only reference to the tensor) will free the memory on the server side.

Let me check with the xla team to see if there is a way to use the buffer while "splitting" a tensor. My feeling is that it would be difficult. We do have the mechanism to aliasing buffer but that is mostly used by inplace operations and operations output and input has the same shape.

**Comment by bdhirsh** [comment]
@JackCaoG yeah, if you think that implementing a `split_copy()` lowerings on the XLA side will be significantly easier than trying to support `split()` directly for now, then I think this will be easier to implement after the functionalization pass.

**Comment by ronghanghu** [comment]
@JackCaoG Thanks again for the explanation!

I tried your exact `test_view.py` snippet above and its output is somehow *not deterministic* for me. On a v3-8 TPU with `tpu-vm-pt-1.10`, its possible print output for this exact snippet include 4 cases:
* case 0: prints `None`
* case 1: prints `(1, 13612258.0, ((1645121786.7991533, 13612258.0),))` 
* case 2: prints `(2, 14260672.0, ((1645121749.394338, 13629909.0), (1645121768.9117424, 630763.0)))` 
* case 3: prints `(3, 20586410.0, ((1645121987.7622552, 19180956.0), (1645121990.714096, 776445.0), (1645122136.998489, 629009.0)))`

and it seems related to whether I'm running this script as a file or pasting them into an interactive Python command line. (When running the file as a script, I most often get case 0. When copy-pasting into the command line, I most often get case 1 or 2 and very rarely case 3). Note that changing `print(met.metric_data('ReleaseDataHandlesTime'))` to the following
```
import gc; gc.collect()
xm.wait_device_ops()
print(met.metric_data('ReleaseDataHandlesTime'))
```
still often gets me to only case 1 or 2.

*I don't think this is a bug though, and I guess it is related to how soon the handles are released. However, this behavior is probably going to make TPU memory debugging more complicated, so I'm reporting it here.*

---

I also hope to report and confirm with you that empirically I find assigning to `t.data=torch.zeros(1, device=device)` also allows me to free a tensor `t`'s storage while preserving the handle of `t`. The main reasons I want to change a tensor's `t.data` rather than directly setting `t = None` are
1) PyTorch's autograd would anyway create a reference to `t` if it is used in subsequent computation so that it can later back-propagate to it (so it's almost impossible to remove all `t`'s references when using auto-grad) and *
2) by keeping the reference handle of `t`, the same tensor `t` can be added to an `nn.Module`'s parameters or passed to an optimizer even if I try to free/edit its storage afterward.

I created a script `test_memory.py` to illustrate this as follows.

```
ronghanghu@t1v-n-9ab95f32-w-0:~$ cat test_memory.py 
import torch
import torch_xla.core.xla_model as xm
import torch_xla.debug.metrics as met

device = xm.xla_device()
buffer_size = 1024**3  # 4 GB for float32
print(
  f"before allocating 4 GB tensor"
  f"\n\tReleaseDataHandlesTime: {met.metric_data('ReleaseDataHandlesTime')}"
  f"\n\tactual TPU memory: {xm.get_memory_info(device)}" 
)

# allocate a large tensor
p = torch.zeros(buffer_size, device=device, dtype=torch.float32, requires_grad=True)
xm.mark_step()
xm.wait_device_ops()
print(
  f"after allocating 4 GB tensor"
  f"\n\tReleaseDataHandlesTime: {met.metric_data('ReleaseDataHandlesTime')}"
  f"\n\tactual TPU memory: {xm.get_memory_info(device)}" 
)

# split the tensor
p1, p2 = p.split([buffer_size // 2, buffer_size // 2])
xm.mark_step()
xm.wait_device_ops()
print(
  f"after splitting 4 GB tensor"
  f"\n\tReleaseDataHandlesTime: {met.metric_data('ReleaseDataHandlesTime')}"
  f"\n\tactual TPU memory: {xm.get_memory_info(device)}" 
)

# attempt to free the memory buffer of p by setting .data to a small tensor
p.data = torch.zeros(1, device=device)
p1.data = torch.zeros(1, device=device)
p2.data = torch.zeros(1, device=device)
# import gc; gc.collect()  # also works for me w/o this garbage collection call
xm.mark_step()
xm.wait_device_ops()
print(
  f"after freeing .data on all tensors"
  f"\n\tReleaseDataHandlesTime: {met.metric_data('ReleaseDataHandlesTime')}"
  f"\n\tactual TPU memory: {xm.get_memory_info(device)}" 
)
```
On a v3-8 TPU, it consistently prints the following for me:
```
ronghanghu@t1v-n-9ab95f32-w-0:~$ python3 test_memory.py 
2022-02-17 19:00:32.658033: E tensorflow/core/framework/op_kernel.cc:1676] OpKernel ('op: "TPURoundRobin" device_type: "CPU"') for unknown op: TPURoundRobin
2022-02-17 19:00:32.658096: E tensorflow/core/framework/op_kernel.cc:1676] OpKernel ('op: "TpuHandleToProtoKey" device_type: "CPU"') for unknown op: TpuHandleToProtoKey
before allocating 4 GB tensor
        ReleaseDataHandlesTime: None
        actual TPU memory: {'kb_free': 16234480, 'kb_total': 16234496}
after allocating 4 GB tensor
        ReleaseDataHandlesTime: None
        actual TPU memory: {'kb_free': 12040128, 'kb_total': 16234496}
after splitting 4 GB tensor
        ReleaseDataHandlesTime: None
        actual TPU memory: {'kb_free': 7845760, 'kb_total': 16234496}
after freeing .data on all tensors
        ReleaseDataHandlesTime: (1, 23041351.0, ((1645124440.1346579, 23041351.0),))
        actual TPU memory: {'kb_free': 16234288, 'kb_total': 16234496}
```
as shown above, empirically the 8 GB memory usage from tensor allocation and splitting is indeed freed after we do `t.data=torch.zeros(1, device=device)` on all the 3 tensors (and I'm getting only 1 in `ReleaseDataHandlesTime`).

**@JackCaoG Could you confirm this behavior of freeing memory via `t.data=torch.zeros(1, device=device)` is indeed support in PyTorch XLA?**

---

Ultimately I'm trying to manage parameter storage for my ZeRO-3 optimizer implementation -- freeing a layer's parameter storage after its forward pass (while having PyTorch autograd tracing them) and reconstructing the parameter storage before the layer's backward pass. On GPUs I can do it as follows
```
import torch

device = torch.device("cuda:0")

buffer_size = 1024**2  # 1 MB for float32, for example
p = torch.ones(buffer_size, device=device, dtype=torch.float32, requires_grad=True)
loss = p.abs().sum()  # any subsequent computation using `p`
# free `p`'s storage after forward pass
p.data.storage().resize_(0)

# do something that would take a lot of memory and requires freeing `p`'s storage
# ...

# reconstruct `p`'s storage before backward pass
p.data.storage().resize_(buffer_size)
p.data.copy_(torch.ones(buffer_size, device=device, dtype=torch.float32))
loss.backward()
assert p.grad is not None and p.grad.numel() == buffer_size
```

Can I accomplish something similar with `p.data=torch.zeros(1, device=device)` in PyTorch XLA?

Thanks again!

**Comment by JackCaoG** [comment]
Let me get back to you tmr

**Comment by ronghanghu** [comment]
On a separate issue -- there seems to be a bug with the implementation of `.view` in PyTorch XLA that causes different behaviors between CPU/GPU and PyTorch XLA. This discrepancy gives **different computation results** when running the same model code on GPUs vs on TPUs (which might be a surprise to users). I found this bug when tracing a model that have different results between GPU and TPU.

In short, the bug is that PyTorch XLA seems to update a tensor's view when it shouldn't (e.g. because the tensor isn't in-place changed). It is illustrated in the following example:

#### CPU (expected behavior)
```
import torch

device = torch.device("cpu")

x1 = torch.zeros(10, device=device)
y1 = x1.view(2, 5)
# this should update y1 to all 1's
x1.data.add_(1.0)

x2 = torch.zeros(10, device=device)
y2 = x2.view(2, 5)
# this should NOT update y1 because x2 now points to a new memory buffer
# while y1 is a view on top of the old memory buffer
x2.data = torch.ones(10, device=device)

print(f"y1: {y1}")
print(f"y2: {y2}")
```
prints
```
y1: tensor([[1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.]])
y2: tensor([[0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.]])
```
Here, only `y1` is updated to 1's because `x.data` is in-place modified, while `y2` is not (and should not be) updated because `x.data` is assigned to a new memory while `y2` is a view on the old memory buffer.

#### TPU (error behavior)

```
import torch
import torch_xla.core.xla_model as xm

device = xm.xla_device()

x1 = torch.zeros(10, device=device)
y1 = x1.view(2, 5)
# this should update y1 to all 1's
x1.data.add_(1.0)

x2 = torch.zeros(10, device=device)
y2 = x2.view(2, 5)
# this should NOT update y1 because x2 now points to a new memory buffer
# while y1 is a view on top of the old memory buffer
x2.data = torch.ones(10, device=device)

xm.mark_step()
print(f"y1: {y1}")
print(f"y2: {y2}")
```
prints
```
y1: tensor([[1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.]], device='xla:1')
y2: tensor([[1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.]], device='xla:1')
```
As can be seen, the TPU behavior diverges from CPU since here `y2` is erroneously updated to 1's even if `x2.data` is not in-place modified. This results in different computation results between CPU/GPU and TPU and may break model training or inference.

> When we try to assign something to t1.data, pt'/xla knows there is a view associated with t1. For the correctness when we update t1's data, we need to update all of the view that is associated with it. It is a long story getting into how pytorch/xla implement view (since we don't really have storage), but we pretty much remembers what kind of reshape operation we performed on each view and replay all of the update after the reshape.

I think to fix this discrepancy and get the same computational outputs between CPU and TPU, it's necessary to distinguish whether t1's data is being in-place updated (such as `x1.data.add_(1.0)`) or not (such as `x1.data = <another tensor>`. Only the former (in-place) case should trigger an update to all of the views associated with it. And I suspect by fixing this problem, the original [shape mismatch error](https://github.com/pytorch/xla/issues/3330#issuecomment-1027443127) in this GitHub issue should also go away.

**Comment by JackCaoG** [comment]
There are quite a fee questions, let me answer then one by one

I think I mislead you about what `ReleaseDataHandlesTime` metric means. It actually means "the number of the times we attempt to release data handles.". The trick here is that each time we can release more than one data handles.

A better way to count the amount of data handles we release (instead of how many bulk of handles we release) is 

```
DestroyDataHandles: {met.counter_value('DestroyDataHandles')
```

If I do this replacement in your example one, I see.

```
before allocating 4 GB tensor
	DestroyDataHandles: None
	actual TPU memory: {'kb_free': 15623788, 'kb_total': 59240548}
Compilation handle 1935859621797462140
Compuation result: 262635955013902
after allocating 4 GB tensor
	DestroyDataHandles: None
	actual TPU memory: {'kb_free': 11379708, 'kb_total': 59240548}
Compilation handle 9126590339074234568
Computation result: 491226942579363
Computation result: 4291254900889496
after splitting 4 GB tensor
	DestroyDataHandles: None
	actual TPU memory: {'kb_free': 7184156, 'kb_total': 59240548}
releasing 262635955013902
releasing 491226942579363
releasing 4291254900889496
Compilation handle 2861382974760993439
Computation result: 1562995218564502
Computation result: 252988570257344
Computation result: 3279929693192351
after freeing .data on all tensors
	DestroyDataHandles: 3
	actual TPU memory: {'kb_free': 15580264, 'kb_total': 59240548}
releasing 252988570257344
releasing 1562995218564502
releasing 3279929693192351
```

I added some debugging messages on my dev branch. 
`Compilation handle` = handle for a compiled graph
`Computation result` = handle of a result of the execution, most likely it is a data handle belong to a tensor.
`releasing` = when we are releasing a handle.

(If you want I can upstream some of these debugging bits but since we are going through a runtime migration maybe in half years, I don't know if you want to learn this much detail about XRT)

**Comment by ronghanghu** [comment]
I see, then I can try out `DestroyDataHandles` to debug my memory usage. 

> (If you want I can upstream some of these debugging bits but since we are going through a runtime migration maybe in half years, I don't know if you want to learn this much detail about XRT)

I think `DestroyDataHandles` will be sufficient for my debugging use case :)

**Comment by JackCaoG** [comment]
Looking at the example 3 I am not sure if it is doing what you are expected.

```
import torch
import torch_xla.core.xla_model as xm
import torch_xla.debug.metrics as met

device = xm.xla_device()
buffer_size = 1024**3 * 3  # 12 GB for float32
print(
  f"before"
  f"\n\tDestroyDataHandles: {met.counter_value('DestroyDataHandles')}"
  f"\n\tactual TPU memory: {xm.get_memory_info(device)}"
)

import pdb; pdb.set_trace()

# allocate one 12 GB tensor `p`
p = torch.zeros(buffer_size, device=device, dtype=torch.float32, requires_grad=True)
loss_p = p.sum()
p.data = torch.zeros(1, device=device)  # free `p` to save memory for `q` below

# allocate another 12 GB tensor `q`
q = torch.ones(buffer_size, device=device, dtype=torch.float32, requires_grad=True)
loss_q = q.sum()
q.data = torch.zeros(1, device=device)  # free `q` to save memory for `loss_p` back-prop below

# back-propagate to `p` (note that `p.data` is not needed in backward pass for `sum`)
loss_p.backward()
p_grad_sum = p.grad.sum()
p.grad = None  # free `p.grad` to save memory for `loss_q` back-prop below

# back-propagate to `q` (note that `q.data` is not needed in backward pass for `sum`)
loss_q.backward()
q_grad_sum = q.grad.sum()

xm.mark_step()
xm.wait_device_ops()
print(
  f"after"
  f"\n\tDestroyDataHandles: {met.counter_value('DestroyDataHandles')}"
  f"\n\tactual TPU memory: {xm.get_memory_info(device)}"
)
print(f"loss_p: {loss_p.item()}, loss_q: {loss_q.item()}, p_grad_sum: {p_grad_sum.item()}, q_grad_sum: {q_grad_sum.item()}")
```
I run it with `XLA_SAVE_TENSORS_FILE="/tmp/ir.txt"` to get a IR text dump and I saw

```
IR {
  %0 = f32[] prim::Constant(), location=<module>@test_memory.py:18, value=0
  %1 = f32[1]{0} aten::expand(%0), location=<module>@test_memory.py:18, size=(1), ROOT=0
  %2 = f32[] prim::Constant(), location=<module>@test_memory.py:16, value=0
  %3 = f32[3221225472]{0} aten::expand(%2), location=<module>@test_memory.py:16, size=(3221225472)
  %4 = f32[] aten::sum(%3), location=<module>@test_memory.py:17, dimensions=(0), keep_reduced_dimensions=0, dtype=6, ROOT=1
  %5 = f32[] prim::Constant(), location=<module>@test_memory.py:23, value=0
  %6 = f32[1]{0} aten::expand(%5), location=<module>@test_memory.py:23, size=(1), ROOT=2
  %7 = f32[] prim::Constant(), location=<module>@test_memory.py:21, value=1
  %8 = f32[3221225472]{0} aten::expand(%7), location=<module>@test_memory.py:21, size=(3221225472)
  %9 = f32[] aten::sum(%8), location=<module>@test_memory.py:22, dimensions=(0), keep_reduced_dimensions=0, dtype=6, ROOT=3
  %10 = f32[] prim::Constant(), location=_make_grads@__init__.py:68, value=1
  %11 = f32[3221225472]{0} aten::expand(%10), size=(3221225472)
  %12 = f32[] aten::sum(%11), location=<module>@test_memory.py:27, dimensions=(0), keep_reduced_dimensions=0, dtype=6, ROOT=4
  %13 = f32[] prim::Constant(), location=_make_grads@__init__.py:68, value=1
  %14 = f32[3221225472]{0} aten::expand(%13), size=(3221225472), ROOT=5
  %15 = f32[] aten::sum(%14), location=<module>@test_memory.py:32, dimensions=(0), keep_reduced_dimensions=0, dtype=6, ROOT=6
}
```

You see that all 4 huge tensor coexist in the executed graph. This is because execution does not happen until you call `xm.mark_step()`. Even if you drop the reference for `p` by setting its data to 0, `loss_p` actually still holds a reference to `p` since it has not materized yet. if you do 

```
print(torch_xla._XLAC._get_xla_tensors_text([loss_p]))
```
before the `mark_step` you will see
```
(Pdb) print(torch_xla._XLAC._get_xla_tensors_text([loss_p]))
IR {
  %0 = f32[] prim::Constant(), location=<module>@test_memory.py:16, value=0
  %1 = f32[3221225472]{0} aten::expand(%0), location=<module>@test_memory.py:16, size=(3221225472)
  %2 = f32[] aten::sum(%1), location=<module>@test_memory.py:17, dimensions=(0), keep_reduced_dimensions=0, dtype=6, ROOT=0
}
```

If you want to achieve described. You will want to add `xm.mark_step()` after each sum.


The reason you code runs and memory seems correct might be because you are using `0` and `1`. These two are special constant to pytorch/xla and we pass then as `constant` instead of `device_data` to the XLA compiler. XLA compiler might be smart enough to not allocating the space for them when it see a sum. You might want to use `randn` or other hardcoded data to better simulate the real use case.

**Comment by JackCaoG** [comment]
Running the above example, I saw
```
-> xm.wait_device_ops()
(Pdb) 
Compuation result: 3432849436068389
Compuation result: 3832412361689119
Compuation result: 3441266025081433
Compuation result: 4410556296210218
Compuation result: 4206526296996723
Compuation result: 275543802580233
Compuation result: 4328585845236716
> /usr/local/google/home/jackcao/Desktop/test/test_memory.py(39)<module>()
-> print(
(Pdb) 
> /usr/local/google/home/jackcao/Desktop/test/test_memory.py(40)<module>()
-> f"after"
(Pdb) 
after
	DestroyDataHandles: None
	actual TPU memory: {'kb_free': 15908592, 'kb_total': 59240548}
> /usr/local/google/home/jackcao/Desktop/test/test_memory.py(44)<module>()
-> print(f"loss_p: {loss_p.item()}, loss_q: {loss_q.item()}, p_grad_sum: {p_grad_sum.item()}, q_grad_sum: {q_grad_sum.item()}")
(Pdb) c
loss_p: 0.0, loss_q: 3221225472.0, p_grad_sum: 3221225472.0, q_grad_sum: 3221225472.0
releasing 3432849436068389
releasing 3832412361689119
releasing 275543802580233
releasing 3441266025081433
releasing 4410556296210218
releasing 4206526296996723
releasing 4328585845236716
```

Pytorch/XLA did not allocate any handle for the 12GB because it is a result of a `expand` of constant and after the `mark_step` `p`, `q`'s value has been updated to 0. However if `p` and `q` are real datas, we actually need to transfer both of them to the HBM before executing the above graph, which will result in a OOM.

**Comment by JackCaoG** [comment]
> On a separate note -- there seems to be a bug in the IR implementation of `.view` in PyTorch XLA that causes different behaviors between CPU/GPU and PyTorch XLA. This discrepancy gives **different computation results** when running the same model code on GPUs vs on TPUs (which might be a surprise to users). I found this bug when tracing a model that have different results between GPU and TPU.
> 
 @ronghanghu Do you mind opening a separate issue for this one? This thread is getting long and might be hard to follow.

**Comment by ronghanghu** [comment]
> @ronghanghu Do you mind opening a separate issue for this one? This thread is getting long and might be hard to follow.

Submitted a new issue to https://github.com/pytorch/xla/issues/3392

**Comment by ronghanghu** [comment]
@JackCaoG Thanks for the detailed explanation here!

> You see that all 4 huge tensor coexist in the executed graph. This is because execution does not happen until you call xm.mark_step()
> If you want to achieve described. You will want to add xm.mark_step() after each sum.
> The reason you code runs and memory seems correct might be because you are using 0 and 1. These two are special constant to pytorch/xla and we pass then as constant instead of device_data to the XLA compiler. XLA compiler might be smart enough to not allocating the space for them when it see a sum. You might want to use randn or other hardcoded data to better simulate the real use case.

That's great to know -- the XLA compiler is smart. I tried out switching to `torch.randn` and indeed got OOM in Example 3 above.

> Even if you drop the reference for p by setting its data to 0, loss_p actually still holds a reference to p since it has not materized yet.
> However if p and q are real datas, we actually need to transfer both of them to the HBM before executing the above graph, which will result in a OOM.

I still didn't understand it. Could you elaborate a bit further on this? In particular, when is it required that all tensors transfer to HBM?

For example, I tried the following case which simulates a neural network with layers that have large internal computation (4 GB from `torch.randn`) but small input and output, which is closer to actual use cases. I could run it even with 10 such layers stacked together. Could you enlighten me on why it doesn't blow up the TPU memory (while Example 3 above does)?

```
import torch
import torch_xla.core.xla_model as xm

device = xm.xla_device()
buffer_size = 1024**3  # 4 GB for float32
num_layers = 10


def layer(input):
    """a layer with large internal computation but small input and output"""
    p = torch.randn(buffer_size, device=device, dtype=torch.float32)
    return input + p.mean()


out = torch.zeros(1, device=device)
for _ in range(num_layers):
    out = layer(out)

xm.mark_step()
xm.wait_device_ops()
print(f"after -- actual TPU memory: {xm.get_memory_info(device)}")
print(f"out: {out.item()}")
```
prints
```
after -- actual TPU memory: {'kb_free': 7835760, 'kb_total': 16234496}
out: -0.00011712321429513395
```
Is the XLA compiler doing something smart here to avoid memory allocation again even if we are using `torch.randn`?

**Comment by JackCaoG** [comment]
Hmm, I need to talk with the xla team to understand the full detail, I can tell you what I know right now.

`torch.tensor(1,2,3,4,5, device='xla:0')` and `torch.randn(5, device='xla:0')` are not the same to the pytorch/xla even they all supposedly take up 20 bytes of device storage.


For the `torch.tensor(1,2,3,4,5, device='xla:0')`, we need to allocate the tensor with 20 byte storage and transfer the data to the runtime before graph can be executed. There is a handle assoicated with this tensor prior to the graph execution.

For the `torch.randn(5, device='xla:0')`, we don't need to allocate the tensor before the graph execution. We only need to transfer a random seed to the XLA runtime(and this happens only once per step, this base random seed will be used to generate subsequence random seeds within this step). The actual random data generation happens during the graph execution instead of prior to the graph execution. There is no handle generated for this `torch.randn` tensor until the graph finished executing.

Here is an example
```
# t1 is a `device_data` with a handle allocated
t1 = torch.tensor([1,2,3,4], device=device)
# assign a new IR to t1, it is no longer a device data but a computation to be executed
t1 += 5 
# execute the pending graph, now t1 gets a new handle which point to a new storage with value ([6,7,8,9]).
# the old storage of ([1,2,3,4]) get garbage collected.
xm.mark_step()
```

Here is the example for randn
```
# we will allocate a handle for the rng seed(4 byte) and t1 will have a IR based on this rng seed
# t1 is not a device_data so no handle assigned to it
t1 = torch.randn(5, device=device)
# update t1's IR
t1 += 5
# execute the pending graph, now t1 gets a new handle. No handle is being destoryed.
xm.mark_step()
```

While I try to figure out what is going on with the randn, is the `randn` more close to your relase case or it will be a `data` that needs to be transfer to the runtime?

**Comment by JackCaoG** [comment]
I guess the question is more of how does xla compiler handles the intermediate tensor and what if total size of the intermediate tensor exceed the HBM. By intermediate I mean this tensor is not the input of the graph (in this case `randn` was generated during the execution, same for `expand`) or the output of the graph(we drop the reference to it so it is also not the output of the graph).

**Comment by ronghanghu** [comment]
@JackCaoG Thanks again for the explanation. The `randn` or intermediate computation is the case I'm encountering in my usage. I don't need the large tensor to be transferred to the runtime (in the case of `data`). Instead, the large tensors will be generated as part of the intermediate computation ops such as concatenation, matrix multiplication, or `xm.all_gather`.

I hope to get a better understanding on how/when the XLA compiler frees an intermediate tensor's storage and what are the conditions to make this happen (e.g. perhaps one condition should be that all references to this tensor are removed and this tensor will no longer be used in any subsequent computation). I'm trying to edit my code to generate an IR graph that allows the XLA compiler to free as many intermediate results as possible in order to train larger models.

My prominent use cases of freeing more intermediate tensors involve 1) ZeRO-3 optimizer implementation where I try to remove a tensor reference via `t.data = torch.zeros(1, device=device)` after forward pass and ask PyTorch's autograd to use another later reconstructed version of `t.data` for its backward pass, and 2) [gradient checkpointing](https://pytorch.org/docs/stable/checkpoint.html) where I similarly want to remove references to intermediate tensors (that would otherwise be referenced and used by autograd) allow the XLA compiler to free them (and recompute them later for the backward pass).

> I guess the question is more of how does xla compiler handles the intermediate tensor and what if total size of the intermediate tensor exceed the HBM.

Yeah, it would be great to understand how the XLA compiler handles it to accomplish memory control for use cases for model scaling such as ZeRO-3 or gradient checkpointing. Thanks again for your help!

**Comment by JackCaoG** [comment]
I checked with the XLA team. I think the conclusion is that XLA compiler reclaim the unneeded buffer is a best effort basis. It is not perfect but has been pretty affective so far. What user needs to do is to make it clear that certain tensor value is not needed after the computation(by dropping the reference to it making it an intermediate value instead of output of the graph exeuction) and XLA compiler will handle the rest. One thing to note is that some of the intermediate tensor might not exist after the optimization pass because of the fused kernel.

If you ever run into a situation that large intermediate tensor caused the OOM, you can dump the HLO by setting `XLA_SAVE_TENSORS_FMT="hlo" XLA_SAVE_TENSORS_FILE="/tmp/ir2.txt"` and I can ask XLA team to see if is there is anything we can do to make it work.


Take way is
1. Drop the reference to unneeded tensors 
2. Let XLA compiler to handle the rest

Let me also come up with an example to explain what is considered to the input and output of the graph.

**Comment by JackCaoG** [comment]
Using your example 5(only difference being that I changed `num_layers` to 5) 

```
import torch
import torch_xla
import torch_xla.core.xla_model as xm
import torch_xla.debug.metrics as met

device = xm.xla_device()
buffer_size = 1024**3  # 4 GB for float32
num_layers = 5


def layer(input):
    """simulate a layer with large internal computation but small input and output"""
    p = torch.randn(buffer_size, device=device, dtype=torch.float32)
    return input + torch.sum(torch.sin(p))

# run the computation
out = torch.zeros(1, device=device)
out_list = []
for _ in range(num_layers):
    out = layer(out)
    out_list.append(out)

import pdb; pdb.set_trace()

xm.mark_step()
xm.wait_device_ops()
print(f"after compute:\n\tDestroyDataHandles: {met.metric_data('DestroyDataHandles')}\n\tactual TPU memory: {xm.get_memory_info(device)}")
print(f"out: {out.item()}")
for idx in range(num_layers):
    print(f"out_list[{idx}]: {out_list[idx]}")
```


with `XLA_SAVE_TENSORS_FMT="hlo" XLA_SAVE_TENSORS_FILE="/tmp/ir2.txt"` I see



```
.....
ENTRY %IrToHlo.351 (p0.2: s64[]) -> (f32[1], f32[1], f32[1], f32[1], f32[1])
....
```

This means this computation takes 1 `s64` input and return 5 `f32` output. That one `s64` is the RNG seed and 5 output are layer output we stored in `out_list`. Everything else is considered intermediate and XLA will take care of them.


In general, a `device_data` that is used to compute the result of another tensor is considered as the input to the graph. The tensor that would get a new value after the graph execution is considered as the `output`  of the graph. We also have a helper function `print(torch_xla._XLAC._xla_tensors_report(1))`

If I call this function right before the `mark_step`, it will print all of the tensor that is live, meaning it has pending computation that needs to be executed. These tensors are likely the output of the graph.

```
(Pdb) print(torch_xla._XLAC._xla_tensors_report(1))
Tensor: id=5, shape=f32[1]{0}, device=CPU:0, ir_nodes=17
Python Frames:
  layer (/usr/local/google/home/jackcao/Desktop/test/test_example5.py:14)
  <module> (/usr/local/google/home/jackcao/Desktop/test/test_example5.py:20)
IR {
  %0 = f32[] prim::Constant(), location=layer@test_example5.py:14, value=1
....
Tensor: id=9, shape=f32[1]{0}, device=CPU:0, ir_nodes=31
....
Tensor: id=13, shape=f32[1]{0}, device=CPU:0, ir_nodes=45
......
Tensor: id=17, shape=f32[1]{0}, device=CPU:0, ir_nodes=59
....
Tensor: id=21, shape=f32[1]{0}, device=CPU:0, ir_nodes=73
```

**Comment by ronghanghu** [comment]
> I think the conclusion is that XLA compiler reclaim the unneeded buffer in a best effort basis. It is not perfect but has been pretty effective so far. What user needs to do is to make it clear that certain tensor value is not needed after the computation(by dropping the reference to it making it an intermediate value instead of output of the graph exeuction) and XLA compiler will handle the rest.

I see. I'll try to free and remove references to the intermediate tensors so that the XLA compiler can free them. 

> If you ever run into a situation that large intermediate tensor caused the OOM, you can dump the HLO by setting XLA_SAVE_TENSORS_FMT="hlo" XLA_SAVE_TENSORS_FILE="/tmp/ir2.txt" and I can ask XLA team to see if is there is anything we can do to make it work.
> In general, a device_data that is used to compute the result of another tensor is considered as the input to the graph. The tensor that would get a new value after the graph execution is considered as the output of the graph. We also have a helper function print(torch_xla._XLAC._xla_tensors_report(1))

The example you showed above is quite helpful. I'll try using these two env variables and torch_xla._XLAC._xla_tensors_report for debugging in case I run into further memory issues.

So the only remaining issue for my use case is https://github.com/pytorch/xla/issues/3392 (it would be also helpful to change `torch.split` to a view op, but `torch.split` is not blocking me and I can get around it by other means). And this one can be closed now.
